### PR TITLE
Docs4

### PR DESCRIPTION
**Pull request by Benedict-Carling**
New attribute to build class

```python
build = bsb.BasicBuild(
        bsb.BasicAssembly(
            "First_Assembly_With_18",
            bsb.BASIC_SEVA_PARTS["v0.1"]["18"],
            bsb.BASIC_BIOLEGIO_LINKERS["v0.1"]["LMP"],
            bsb.BASIC_CDS_PARTS["v0.1"]["sfGFP"],
            bsb.BASIC_BIOLEGIO_LINKERS["v0.1"]["LMS"],
        ),
        bsb.BasicAssembly(
            "Second_Assembly_With_26",
            bsb.BASIC_SEVA_PARTS["v0.1"]["26"],
            bsb.BASIC_BIOLEGIO_LINKERS["v0.1"]["LMP"],
            bsb.BASIC_CDS_PARTS["v0.1"]["sfGFP"],
            bsb.BASIC_BIOLEGIO_LINKERS["v0.1"]["LMS"],
        ),
    )

build.export_pdf()
```

Creates the following pdf
[pdf_10-02-2021_13.23.26.pdf](https://github.com/LondonBiofoundry/basicsynbio/files/5958456/pdf_10-02-2021_13.23.26.pdf)

Unit tests are added to ensure expected size for the build above, contains assembly id's `Second_Assembly_With_26`, contains parts names `BASIC_sfGFP_CDS` and contains expected template strings such as `materials`.

### CODE REVIEW

**Comment by hainesm6** [comment]
Additional notes on current pdf:
- Need to have a master mix for the clip reactions rather than setting them up with individual components. *Write formulae to calculate constant to multiply table values by*.

**Comment by hainesm6** [comment]
@Benedict-Carling could you also update the documentation.

**Comment by hainesm6** [comment]
@Benedict-Carling seperate out this method into it's own module to accept a `BasicBuild` object. That along with the master mix table should be it 🍾

**Comment by hainesm6** [comment]
hey @Benedict-Carling! I've added a csv file (last commit) describing clip reaction master mix components. An array containing the total volume of each component should be calculated as follows:

```python
dead_volume = 1 + total_clips//20
array = component_array*(total_clips + dead_volume)
```

The `total_clips` variable can be caluclated from the `BasicBuild.clips_data` attribute. I would say each clip should be used in up to 28 assemblies according to the instructions but we should have this as a default argument in function/method. Hope that's all alright but feel free to let me know if anything needs explaining👍

**Comment by Benedict-Carling** [comment]
Hi @hainesm6 This pull request meets requirements again with new dynamic volume master mix table 🎉, the pdf for the build In the original pull request comment now outputs
[pfd_instructions.pdf](https://github.com/LondonBiofoundry/basicsynbio/files/6024304/pfd_instructions.pdf)

File: basicsynbio/functions/pdf_instructions.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+            working directory with a time stamped name, output csvs is created.
+
+    Returns:
+        str: filepath of created pdf
+    """
+
+    import pandas as pd
+    import math
+
+    def calculate_clip_num(assemblies: list, assemblies_per_clip: int = 28):
```
Comment (hainesm6) [inline basicsynbio/functions/pdf_instructions.py:52]: Can we move/duplicate the `assemblies_per_clip` argument to the `pdf_instructions` function? Users can then specify how many assemblies each clip should be used in.

File: basicsynbio/functions/pdf_instructions.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+                Paragraph(
+                    "Place Falcon plate on magnetic stand and wait for rings to form and solution to clear.",
+                    styles["BodyText"],
+                ),
+                Paragraph(
+                    "Remove the solution with a 200 μl pipette tip from the centre of each well.",
+                    styles["BodyText"],
+                ),
+                Paragraph(
+                    "Add 190 μl 70'%' EtOH to each well and wait 30 s. ",
```
Comment (hainesm6) [inline basicsynbio/functions/pdf_instructions.py:278]: Can we remove quotation marks around the % sign? I think there's a few other instances of this.

File: basicsynbio/functions/pdf_instructions.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    )
+    elems.append(
+        ListFlowable(
+            [
+                Paragraph(
+                    "Add 54 μl of magnetic beads into 96 well Falcon plate (one well per BASIC reaction) and add the 30 μl BASIC linker ligation from the PCR machine step, mix by pipetting 10 times. ",
+                    styles["BodyText"],
+                ),
+                Paragraph(
+                    "Wait 5min to allow DNA binding to magbeads. ",
```
Comment (hainesm6) [inline basicsynbio/functions/pdf_instructions.py:266]: "5 min"

File: basicsynbio/functions/pdf_instructions.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+                    styles["BodyText"],
+                ),
+                Paragraph(
+                    "Place Falcon plate on magnetic stand and wait for rings to form and solution to clear.",
+                    styles["BodyText"],
+                ),
+                Paragraph(
+                    "Remove the solution with a 200 μl pipette tip from the centre of each well.",
```
Comment (hainesm6) [inline basicsynbio/functions/pdf_instructions.py:274]: "Aspirate most of the solution with a 200 µL pipette set to 80 µL."

File: basicsynbio/functions/pdf_instructions.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+                    "Remove Falcon plate from magnet and resuspend magbeads in 32 μl dH20. ",
+                    styles["BodyText"],
+                ),
+                Paragraph("Wait 1 min for DNA to elute. ", styles["BodyText"]),
+                Paragraph(
+                    "Place Falcon plate back on magnetic stand and allow ring to form and solution to clear. ",
+                    styles["BodyText"],
+                ),
+                Paragraph(
+                    "Pipette 30 μl of H20 with eluted DNA into fresh 1.5 ml Eppendorf tube for direct use in assembly or storage at -20°C for up to 1 month. ",
```
Comment (hainesm6) [inline basicsynbio/functions/pdf_instructions.py:304]: "Transfer 30 μl of purified clip reaction into a clean microcentrifuge tube or well and store at at -20°C for up to 1 month."

File: basicsynbio/functions/pdf_instructions.py [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+            styles["BodyText"],
+        )
+    )
+    elems.append(Spacer(1, 0.4 * cm))
+    elems.append(
+        Table(ASSEMBLIES_DATA, colWidths=[5.2 * cm, 5.2 * cm, 5.2 * cm], style=style)
+    )
+    elems.append(
+        Paragraph(
+            "For each BASIC assembly (example here with 3 parts) combine parts with buffer in a PCR tube: ",
```
Comment (hainesm6) [inline basicsynbio/functions/pdf_instructions.py:329]: "For each BASIC assembly, combine the required purified clip reactions in a final volume of 10 μl in 1x Assembly or NEB CutSmart buffer. Below gives an example for a 3-part assembly:"

File: basicsynbio/utils/reportLab.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    ["20", "1 min", ""],
+    ["37", "5 min", ""],
+]
+
+ASSEMBLY_REACTION_3 = [
+    ["Reagent", "Volume "],
+    ["dH20", "2 μl"],
+    ["10x Assembly Buffer (or 10x NEB CutSmart)", "1 μl"],
+    [
+        "Each Linker Ligated Basic Part (clip) within the assembly",
```
Comment (hainesm6) [inline basicsynbio/utils/reportLab.py:113]: Each purified clip reaction required for the assembly

File: basicsynbio/utils/reportLab.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+TEMPERATURE_PROFILE_1 = [
+    ["Temperature (°C)", "Time", ""],
+    ["37", "2 min", "x20 cycles"],
+    ["20", "1 min", ""],
+    ["37", "5 min", ""],
+]
+
+ASSEMBLY_REACTION_3 = [
+    ["Reagent", "Volume "],
+    ["dH20", "2 μl"],
```
Comment (hainesm6) [inline basicsynbio/utils/reportLab.py:110]: [Delete this row for dH2O as it's not necessary]

File: basicsynbio/functions/pdf_instructions.py [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+                Paragraph(
+                    "Chemically competent cells are stored at -80°C. ",
+                    styles["BodyText"],
+                ),
+                Paragraph(
+                    "Thaw competent cells on ice (takes 5-10 min); 50 μl per BASIC assembly to be transformed. ",
+                    styles["BodyText"],
+                ),
+                Paragraph(
+                    "Cool 5 μl of BASIC DNA assembly in 1.5 ml Eppendorf tube on ice. ",
```
Comment (hainesm6) [inline basicsynbio/functions/pdf_instructions.py:380]: Replace "Eppendorf" with "microcentrifuge"

**Review by hainesm6 — APPROVED** [review]

**Comment by Benedict-Carling** [comment]
Implements new changes suggested today! resulting output is now:

[pdf_23-02-2021_18.44.24.pdf](https://github.com/LondonBiofoundry/basicsynbio/files/6031178/pdf_23-02-2021_18.44.24.pdf)

**Pull request merged**